src/Microsoft.AspNet.JsonPatch/JsonPatchDocument.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNet.JsonPatch.Adapters;
+using Microsoft.AspNet.JsonPatch.Converters;
+using Microsoft.AspNet.JsonPatch.Exceptions;
+using Microsoft.AspNet.JsonPatch.Helpers;
+using Microsoft.AspNet.JsonPatch.Operations;
+using Microsoft.Framework.Internal;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.AspNet.JsonPatch
+{
+    [JsonConverter(typeof(JsonPatchDocumentConverter))]
+    public class JsonPatchDocument : IJsonPatchDocument
+    {
+
+        public List<Operation> Operations { get; private set; }
+
+        [JsonIgnore]
+        public IContractResolver ContractResolver { get; set; }
+
+        /// <summary>
+        /// Action for logging <see cref="JsonPatchError"/>.
+        /// </summary>
+        public Action<JsonPatchError> LogErrorAction { get; }
+
+
+        public JsonPatchDocument()
+        {
+            Operations = new List<Operation>();
+            ContractResolver = new DefaultContractResolver();             
+        }
+
+
+        public JsonPatchDocument(Action<JsonPatchError> logErrorAction)
+            : this()
+        {
+            LogErrorAction = LogErrorAction;
+        }
+
+        // Create from list of operations - logErrorAction isn't required in this case, as
+        // logging errors through that specific logErrorAction happens when the PatchDocument
+        // is created, not after deserialization.
+        public JsonPatchDocument([NotNull] List<Operation> operations, [NotNull]  IContractResolver contractResolver)
+        {
+            Operations = operations;
+            ContractResolver = contractResolver;
+        }
+
+
+        public JsonPatchDocument Add([NotNull] string path, object value)
+        {
+            var checkPathResult = PathHelpers.CheckPath(path);
+            if (!checkPathResult.IsCorrectlyFormedPath)
+            {
+                LogError(new JsonPatchError(
```

Comment:
Not used for anything else - didn't realize it should only be used during deserialization; I'll remove the logError function from JsonPatchDocument